### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by HiromuHota**
I created a Dockerfile and docker-compose.yml for the sake of ease of deployment.
Please take a look and merge it if it is useful for other users.
Let me know if you want commits to be squashed before being merged.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    snorkel-metal==0.2.4
+
+RUN python -m spacy download en
+RUN conda install -y -c conda-forge ipywidgets
+
+USER $NB_UID
+
+COPY --chown=jovyan:users hardware hardware
+COPY --chown=jovyan:users hardware_image hardware_image
+COPY --chown=jovyan:users intro intro
```
Comment (lukehsiao) [inline Dockerfile:29]: Having never used Docker, I'm not qualified to do a code review. But, I do notice several places that seem very specific to a particular user like `jovyan` being used here and in `docker-compose.yml`. I'm not sure if this is necessary/intended or if there are more generalizable ways to do this.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+RUN pip install torch==0.4.1
+RUN pip install \
+    fonduer==0.3.4 \
+    matplotlib \
+    snorkel-metal==0.2.4
+
+RUN python -m spacy download en
+RUN conda install -y -c conda-forge ipywidgets
+
+USER $NB_UID
```
Comment (lukehsiao) [inline Dockerfile:25]: Here's another example of a variable where I'm not sure where it comes from or how it is set. Will this work for all users?

**Review by lukehsiao — COMMENTED** [review]
I'm not familiar enough with Docker to give a good review. @SenWu any additional comments? 

Overal, I trust Hiromu to create a good Docker set-up, even if I don't understand what's happening in this code.

File: Dockerfile (hunk 1 above)
Comment (HiromuHota) [reply Dockerfile:29]: `jovyan` is originally defined by https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile#L10 in the base Docker image.
Note the Docker image inheritance is the current Docker image (say `hiromuhota/fonduer-tutorials`) -> `jupyter/minimal-notebook` -> `jupyter/base-notebook` -> `ubuntu`.
In `jupyter/base-notebook`, the jupyter server is ran by `jovyan`. That's why I changed the owner of these folder/files as `jovyan`.

File: Dockerfile (hunk 2 above)
Comment (HiromuHota) [reply Dockerfile:25]: The direct answer to your question is no it won't work for all **linux** users within a Docker container, but this is how Docker works and should be ok for tutorials.
This Docker image makes it easy to jump start their Fonduer tutorials and is never meant to be used in production.
Well, end users have to have Docker installed but this is increasingly becoming popular.

Once a Docker container is deployed, the jupyter server is ran by `jovyan`.
As you know, jupyter itself is a web app and can serve multiple **human** users, which is good enough for tutorials.

Comment (HiromuHota) [reply Dockerfile:25]: Having said, you don't want multiple human users to use a single Fonduer instance.
In practice, each user has to have a dedicated Fonduer instance for these tutorials, which could be deployed to their local PC or on the cloud.

File: Dockerfile (hunk 1 above)
Comment (lukehsiao) [reply Dockerfile:29]: Ah, I see. Makes sense.

**Review by lukehsiao — APPROVED** [review]
Given Hiromu's responses, this seems fine to me.

**Pull request merged**